frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -0,0 +1,1452 @@
+			+ IsType<<Self as frame_system::Config>::Event>;
+
+		/// Currency type.
+		type Currency: ReservableCurrency<Self::AccountId> + Currency<Self::AccountId>;
+
+		/// Duration of the unsigned phase.
+		#[pallet::constant]
+		type UnsignedPhase: Get<Self::BlockNumber>;
+		/// Duration of the signed phase.
+		#[pallet::constant]
+		type SignedPhase: Get<Self::BlockNumber>;
+
+		/// The minimum amount of improvement to the solution score that defines a solution as
+		/// "better" (in any phase).
+		#[pallet::constant]
+		type SolutionImprovementThreshold: Get<Perbill>;
+
+		/// The priority of the unsigned transaction submitted in the unsigned-phase
+		type MinerTxPriority: Get<TransactionPriority>;
+		/// Maximum number of iteration of balancing that will be executed in the embedded miner of
+		/// the pallet.
+		type MinerMaxIterations: Get<u32>;
+		/// Maximum weight that the miner should consume.
+		///
+		/// The miner will ensure that the total weight of the unsigned solution will not exceed
+		/// this values, based on [`WeightInfo::submit_unsigned`].
+		type MinerMaxWeight: Get<Weight>;
+
+		/// Something that will provide the election data.
+		type DataProvider: ElectionDataProvider<Self::AccountId, Self::BlockNumber>;
+
+		/// The compact solution type
+		type CompactSolution: codec::Codec
+			+ Default
+			+ PartialEq
+			+ Eq
+			+ Clone
+			+ sp_std::fmt::Debug
+			+ CompactSolution;
+
+		/// Accuracy used for fallback on-chain election.
+		type OnChainAccuracy: PerThing128;
+
+		/// Configuration for the fallback
+		type Fallback: Get<FallbackStrategy>;
+
+		/// The configuration of benchmarking.
+		type BenchmarkingConfig: BenchmarkingConfig;
+
+		/// The weight of the pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(now: T::BlockNumber) -> Weight {
+			let next_election = T::DataProvider::next_election_prediction(now).max(now);
+
+			let signed_deadline = T::SignedPhase::get() + T::UnsignedPhase::get();
+			let unsigned_deadline = T::UnsignedPhase::get();
+
+			let remaining = next_election - now;
+			let current_phase = Self::current_phase();
+
+			match current_phase {
+				Phase::Off if remaining <= signed_deadline && remaining > unsigned_deadline => {
+					Self::on_initialize_open_signed();
+					log!(info, "Starting signed phase at #{:?} , round {}.", now, Self::round());
+					T::WeightInfo::on_initialize_open_signed()
+				}
+				Phase::Signed | Phase::Off
+					if remaining <= unsigned_deadline && remaining > 0u32.into() =>
+				{
+					let (need_snapshot, enabled, additional) = if current_phase == Phase::Signed {
+						// followed by a signed phase: close the signed phase, no need for snapshot.
+						// NOTE: SIGNED_PHASE
+						(false, true, Weight::zero())
+					} else {
+						// no signed phase: create a new snapshot, definitely `enable` the unsigned
+						// phase.
+						(true, true, Weight::zero())
+					};
```

Comment:
additional is always weight 0?

Reply:
This will be filled in the signed phase. Anywhere that you see `// NOTE: SIGNED_PHASE`, best to move the reviw to that PR.

Reply:
Yes, in this case we don't do anything. In the signed PR, it should be the case that the weight of closing the signed phase must be passed as the last argument as `additional`. Not sure if that PR is updated to include this, but this is how it is meant to be.

Reply:
note that the signed PR doesn't change this code, but I guess it is a miss